tokio/src/runtime/scheduler/multi_thread/worker.rs

```diff
@@ -793,6 +849,143 @@ impl Context {
             self.defer.defer(waker);
         }
     }
+
+    cfg_time! {
+        /// Maintain local timers before parking the resource driver.
+        ///
+        /// * Remove cancelled timers from the local timer wheel.
+        /// * Register remote timers to the local timer wheel.
+        /// * Adjust the park duration based on
+        ///   * the next timer expiration time.
+        ///   * whether auto-advancing is required (feature = "test-util").
+        ///
+        /// # Returns
+        ///
+        /// `(Box<Core>, park_duration, auto_advance_duration)`
+        fn maintain_local_timers_before_parking(
+            &self,core: Box<Core>,
+            park_duration: Option<Duration>
+        ) -> MaintainLocalTimer {
+            let handle = &self.worker.handle;
+            let (core, (should_yield, next_timer)) =
+                self.enter_with_time_context(core, |time_cx| {
+                    util::time::remove_cancelled_timers(&mut time_cx.wheel, &mut time_cx.canc_rx);
+                    let should_yield = util::time::insert_inject_timers(
+                        &mut time_cx.wheel,
+                        &time_cx.canc_tx,
+                        handle.take_remote_timers(),
+                    );
```

Comment:
For cancellation, we probably also have to move timers to linked list, and cancel after releasing `RefMut`. (I think.)